### PR TITLE
[Flow EVM] Adding metrics to the EVM

### PR DESCRIPTION
**Pull request by ramtinms**
This PR adds metrics to the EVM. 

ps. The majority of the code changes are due to the mocks getting updated.

### CODE REVIEW

File: module/metrics/execution.go [hunk 1]
```diff
@@ -959,6 +1019,35 @@ func (ec *ExecutionCollector) RuntimeTransactionProgramsCacheHit() {
 	ec.programsCacheHit.Inc()
 }
 
+func (ec *ExecutionCollector) SetNumberOfDeployedCOAs(count uint64) {
+	ec.numberOfDeployedCOAs.Set(float64(count))
+}
+
+func (ec *ExecutionCollector) EVMTransactionExecuted(
```
Comment (janezpodhostnik) [inline module/metrics/execution.go:1026]: from this setup there will be no way to tell if there are more direct call failures than non-direct call failures. Is that something we need?

File: module/metrics/execution.go [hunk 2]
```diff
@@ -959,6 +1019,35 @@ func (ec *ExecutionCollector) RuntimeTransactionProgramsCacheHit() {
...
+	}
+	ec.evmTransactionGasUsed.Observe(float64(gasUsed))
+}
+
+func (ec *ExecutionCollector) EVMBlockExecuted(
+	txCount int,
+	totalGasUsed uint64,
+	totalSupplyInFlow float64,
+) {
+	ec.evmBlockTxCount.Observe(float64(totalGasUsed))
```
Comment (janezpodhostnik) [inline module/metrics/execution.go:1046]:
```suggestion
	ec.evmBlockTxCount.Observe(float64(txCount))
```

File: fvm/evm/handler/handler.go [hunk 3]
```diff
@@ -267,6 +276,13 @@ func (h *ContractHandler) commitBlockProposal() error {
 		return err
 	}
 
+	// report metrics
+	h.backend.EVMBlockExecuted(
+		len(bp.TxHashes),
+		bp.TotalGasUsed,
+		types.UnsafeCastOfBalanceToFloat64(bp.TotalSupply),
```
Comment (janezpodhostnik) [inline fvm/evm/handler/handler.go:283]: should the cast be done by the metrics collector?

**Review by janezpodhostnik — APPROVED** [review]

File: fvm/evm/handler/handler.go (hunk 3 above)
Comment (ramtinms) [reply fvm/evm/handler/handler.go:283]: I first put it there but then it had to have dependency to Cadence internal logic

File: module/metrics/execution.go (hunk 2 above)
Comment (ramtinms) [reply module/metrics/execution.go:1046]: nice find. fixed

File: module/metrics/execution.go (hunk 1 above)
Comment (ramtinms) [reply module/metrics/execution.go:1026]: good idea, I can add that as well.

Comment (ramtinms) [reply module/metrics/execution.go:1026]: added

**Review by janezpodhostnik — APPROVED** [review]

**Review by sideninja — APPROVED** [review]

**Pull request merged**